File: src/fp11/closure.c
/* Closure implementation in C */

#include <stdio.h>

typedef struct {
  int (*fp)(int, int, int, int);
  int fv1; int fv2; } Closure_II_II_I;

int apply_II_II_I(int v1, int v2, Closure_II_II_I *closure) {
  int (*fp)(int, int, int, int) = closure->fp;
  int fv1 = closure->fv1, fv2 = closure->fv2;
  return fp(v1, v2, fv1, fv2);
}

int f_in_c_style(int x, int y, int a, int b) {
  return a * x + b * y;
}

int main() {
  const int a = 3, b = 4;
  Closure_II_II_I closure_f = { f_in_c_style, a, b };
  printf("%d\n", apply_II_II_I(5, 6, &closure_f));
}

File: src/fp11/closure.ml
(* FP-style function application *)

let a = 3 in
let b = 4 in
let rec f x y =
  a * x + b * y in
Printf.printf "%d\n" (f 5 6);;

(* C-style emulation using the closure technique *)

(* Application of a function closure that takes
 *  - two arguments explicitly passed as function parameters (arg1, arg2) and
 *  - two arguments that correspond to free variables.  They are saved in the closure structure at the function calling site
 **)
let apply_2arg_2fv arg1 arg2 closure =
  let a_function_implementation, (fv1, fv2) = closure in
  a_function_implementation arg1 arg2 fv1 fv2 in

let rec f_implementation x y a b =
  a * x + b * y in

let a = 3 in
let b = 4 in
let closure = (f_implementation, (a, b)) in
Printf.printf "%d\n" (apply_2arg_2fv 5 6 closure)
